File: parser/go/lexspec_test.go
package parser_test

// The Go half of parser/spec/lex.tsv. The TypeScript runner has its own
// reader and renderer: keeping both implementations independent prevents a
// shared harness bug from making two different token streams look equal.

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/boru-lang/boru/parser/go"
)

const lexSpecRowCount = 30

type lexSpecRow struct {
	line     int
	src      string
	status   string
	expected string
}

type lexSpecToken struct {
	Name string `json:"name"`
	Src  string `json:"src"`
	SI   int    `json:"si"`
}

func readLexSpec(t *testing.T) []lexSpecRow {
	t.Helper()
	const name = "lex.tsv"
	path := filepath.Join("..", "spec", name)
	f, err := os.Open(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	defer f.Close()

	seen := make(map[string]int)
	var rows []lexSpecRow
	sc := bufio.NewScanner(f)
	for n := 1; sc.Scan(); n++ {
		line := sc.Text()
		if line == "" || (strings.HasPrefix(line, "#") && !strings.Contains(line, "\t")) {
			continue
		}
		parts := strings.Split(line, "\t")
		if len(parts) != 3 {
			t.Fatalf("%s:%d: need exactly 3 tab-separated columns (source, status, tokens), got %d", name, n, len(parts))
		}
		src := decodeSpecEscapes(parts[0])
		if first, exists := seen[src]; exists {
			t.Fatalf("%s:%d: duplicate source (first at line %d): %q", name, n, first, src)
		}
		seen[src] = n
		status := decodeSpecEscapes(parts[1])
		if status != "OK" && status != "ERR" {
			t.Fatalf("%s:%d: status %q is not OK or ERR", name, n, status)
		}
		expected := decodeSpecEscapes(parts[2])
		validateLexSpecTokens(t, name, n, expected)
		rows = append(rows, lexSpecRow{line: n, src: src, status: status, expected: expected})
	}
	if err := sc.Err(); err != nil {
		t.Fatalf("scan %s: %v", path, err)
	}
	if len(rows) != lexSpecRowCount {
		t.Fatalf("%s: got %d data rows, want exact ratchet %d", path, len(rows), lexSpecRowCount)
	}
	return rows
}

func validateLexSpecTokens(t *testing.T, name string, line int, expected string) {
	t.Helper()
	var tokens []lexSpecToken
	dec := json.NewDecoder(strings.NewReader(expected))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&tokens); err != nil {
		t.Fatalf("%s:%d: tokens are not a canonical token array: %v", name, line, err)
	}
	if _, err := dec.Token(); err != io.EOF {
		t.Fatalf("%s:%d: trailing data after token array", name, line)
	}
	for i, token := range tokens {
		if !strings.HasPrefix(token.Name, "#") || token.SI < 0 {
			t.Fatalf("%s:%d: invalid token %d: %+v", name, line, i, token)
		}
	}
	canonical, err := marshalLexSpecTokens(tokens)
	if err != nil {
		t.Fatalf("%s:%d: marshal expected tokens: %v", name, line, err)
	}
	if canonical != expected {
		t.Fatalf("%s:%d: token JSON is not canonical; want %s", name, line, canonical)
	}
}

// encoding/json's package helper HTML-escapes `>` while JSON.stringify does
// not. Disable that optional transport hardening so both independent token
// renderers use the same canonical source spelling (notably for `=>`).
func marshalLexSpecTokens(tokens []lexSpecToken) (string, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(tokens); err != nil {
		return "", err
	}
	return strings.TrimSuffix(buf.String(), "\n"), nil
}

func renderLexSpec(tokens []parser.LexToken) string {
	out := make([]lexSpecToken, len(tokens))
	for i, token := range tokens {
		out[i] = lexSpecToken{Name: token.Name, Src: token.Src, SI: token.SI}
	}
	rendered, err := marshalLexSpecTokens(out)
	if err != nil {
		panic(fmt.Sprintf("render lex tokens: %v", err))
	}
	return rendered
}

func TestParserSpecLex(t *testing.T) {
	for _, row := range readLexSpec(t) {
		row := row
		t.Run(fmt.Sprintf("line-%d", row.line), func(t *testing.T) {
			tokens, ok := parser.LexTokens(row.src)
			status := "ERR"
			if ok {
				status = "OK"
			}
			if status != row.status {
				t.Errorf("lex.tsv:%d: %q status: want %s, got %s", row.line, row.src, row.status, status)
			}
			if got := renderLexSpec(tokens); got != row.expected {
				t.Errorf("lex.tsv:%d: %q tokens:\n  want: %s\n  got : %s", row.line, row.src, row.expected, got)
			}
			// Every returned raw token must begin at its advertised UTF-8 byte
			// offset. This assertion is independent of the checked-in render.
			for _, token := range tokens {
				if token.SI < 0 || token.SI > len(row.src) ||
					!bytes.HasPrefix([]byte(row.src)[token.SI:], []byte(token.Src)) {
					t.Errorf("lex.tsv:%d: token %+v does not index source bytes", row.line, token)
				}
			}
		})
	}
}

File: parser/ts/src/lexspec.test.ts
// The TypeScript half of parser/spec/lex.tsv. Its strict reader and renderer
// intentionally share no implementation with parser/go/lexspec_test.go.

import { strict as assert } from 'node:assert'
import * as fs from 'node:fs'
import * as path from 'node:path'
import { describe, it } from 'node:test'
import { fileURLToPath } from 'node:url'

import { lexTokens, type LexToken } from './index.ts'

const SPEC_DIR = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..', '..', 'spec')
const LEX_SPEC_ROW_COUNT = 34

interface LexSpecRow {
  line: number
  src: string
  status: 'OK' | 'ERR'
  expected: string
}

function decodeLexSpecEscapes(source: string): string {
  let decoded = ''
  for (let i = 0; i < source.length; i++) {
    if ('\\' === source[i] && i + 1 < source.length) {
      const next = source[i + 1]
      if ('n' === next) {
        decoded += '\n'
        i++
        continue
      }
      if ('t' === next) {
        decoded += '\t'
        i++
        continue
      }
      if ('\\' === next) {
        decoded += '\\'
        i++
        continue
      }
    }
    decoded += source[i]
  }
  return decoded
}

function validateLexSpecTokens(name: string, line: number, expected: string): void {
  let value: unknown
  assert.doesNotThrow(() => {
    value = JSON.parse(expected)
  }, `${name}:${line}: tokens must be JSON`)
  assert.ok(Array.isArray(value), `${name}:${line}: tokens must be an array`)
  for (const [index, token] of value.entries()) {
    assert.ok(null !== token && 'object' === typeof token && !Array.isArray(token), `${name}:${line}: token ${index} must be an object`)
    const record = token as Record<string, unknown>
    assert.deepEqual(Object.keys(record), ['name', 'src', 'si'], `${name}:${line}: token ${index} fields`)
    assert.ok('string' === typeof record.name && record.name.startsWith('#'), `${name}:${line}: token ${index} name`)
    assert.equal(typeof record.src, 'string', `${name}:${line}: token ${index} source`)
    assert.ok(Number.isSafeInteger(record.si) && (record.si as number) >= 0, `${name}:${line}: token ${index} offset`)
  }
  assert.equal(JSON.stringify(value), expected, `${name}:${line}: token JSON must be canonical`)
}

function readLexSpec(): LexSpecRow[] {
  const name = 'lex.tsv'
  const lines = fs.readFileSync(path.join(SPEC_DIR, name), 'utf8').split('\n')
  const rows: LexSpecRow[] = []
  const seen = new Map<string, number>()
  for (let index = 0; index < lines.length; index++) {
    const line = lines[index]!
    if ('' === line || (line.startsWith('#') && !line.includes('\t'))) continue
    const parts = line.split('\t')
    assert.equal(parts.length, 3, `${name}:${index + 1}: need exactly 3 tab-separated columns (source, status, tokens)`)
    const src = decodeLexSpecEscapes(parts[0]!)
    assert.equal(seen.has(src), false, `${name}:${index + 1}: duplicate source (first at line ${seen.get(src)}): ${JSON.stringify(src)}`)
    seen.set(src, index + 1)
    const status = decodeLexSpecEscapes(parts[1]!)
    assert.ok('OK' === status || 'ERR' === status, `${name}:${index + 1}: status must be OK or ERR`)
    const expected = decodeLexSpecEscapes(parts[2]!)
    validateLexSpecTokens(name, index + 1, expected)
    rows.push({ line: index + 1, src, status, expected })
  }
  assert.equal(rows.length, LEX_SPEC_ROW_COUNT, `${name}: exact row-count ratchet`)
  return rows
}

function renderLexSpec(tokens: LexToken[]): string {
  return JSON.stringify(tokens.map(({ name, src, si }) => ({ name, src, si })))
}

describe('parser spec — lex.tsv', () => {
  for (const row of readLexSpec()) {
    it(`${row.line}: ${JSON.stringify(row.src)}`, () => {
      const result = lexTokens(row.src)
      assert.equal(result.ok ? 'OK' : 'ERR', row.status)
      assert.equal(renderLexSpec(result.tokens), row.expected)

      const sourceBytes = Buffer.from(row.src)
      for (const token of result.tokens) {
        const rawBytes = Buffer.from(token.src)
        assert.equal(sourceBytes.subarray(token.si, token.si + rawBytes.length).equals(rawBytes), true)
      }
    })
  }
})
